### PR TITLE
Orb author FAQ page

### PR DESCRIPTION
**Pull request by KyleTryon**
# Description
Proposal to add FAQ section for orb authors. This content has not yet been reviewed by the rest of the CPE team, requesting reviews via this PR which may result in changes. 

Do not merge until all reviewers have signed off.

# Reviewers:
Please review the included content and discuss any thoughts below.

# Reasons
To mirror our existing FAQ for users, this is a similar help doc for authors. This page also compliments our best practices doc.

### CODE REVIEW

File: jekyll/_cci2/orb-author-faq.md [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+### Supported Programming Languages
+
+* Question: What programming language can I write my orb in?
+* Answer: Orbs are packages of [CircleCI YAML config]({{site.baseurl}}/2.0/configuration-reference/) language. CircleCI config may contain commands that are issued against a given [Executor]({{site.baseurl}}/2.0/executor-intro/).
+
+CircleCI orbs package [CircleCI reusable config]({{site.baseurl}}/2.0/reusing-config/), such as [Commands]({{site.baseurl}}/2.0/reusing-config/#authoring-reusable-commands), which can execute within a given [Executor]({{site.baseurl}}/2.0/executor-intro/) defined by either the user if using a command within a custom job, or by the orb author if using a [Reusable Job]({{site.baseurl}}/2.0/orb-author-intro/#jobs). The environment within which your logic is running may influence your language descisions.
+
+**Bash**
+
+Bash is the preferred language as it is most commonly available among all available executors. Bash can be (and should) easily written directly using the native [run]({{site.baseurl}}/2.0/configuration-reference/#run) command. The default shell MacOS and Linx will be Bash.
```
Comment (felicianotech) [inline jekyll/_cci2/orb-author-faq.md]: `Linx` -> `Linux`

More importantly, we should be more specific on behavior I think. By default, the "language" that is used is the default shell for the OS in the executor. This isn't always Bash.

File: jekyll/_cci2/orb-author-faq.md [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+---
+
+This document describes various questions and technical issues that you may find helpful when authoring orbs.
+
+* TOC
+{:toc}
+
+
+
+### Errors Claiming Namespace Or Publishing Orbs
```
[the remark this answers is not in the document]
Comment (KyleTryon) [reply jekyll/_cci2/orb-author-faq.md]: Agreed. Copied this from the orbs-faq page which will need to be updated as well. Including this fix in my next commit.

File: jekyll/_cci2/orb-author-faq.md [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+
+
+## Errors Claiming Namespace Or Publishing Orbs
+
+* Question: I recieve an error when attempting to claim a namespace or publish a production orb.
+
+* Answer: You may not be an organization owner/admin.
+
+Organizations may only claim a single namespace. In order to claim a namespace for an organization the authenticating user must possess owner/admin privledges over the organization.
```
Comment (lokst) [inline jekyll/_cci2/orb-author-faq.md]: Should it be "privileges"?

File: jekyll/_cci2/orb-author-faq.md [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+* Answer: Utilize the "env_var_name" parameter type for the API key parameter. This parameter type will only accept POSIX valid environment variable name strings as valid. In the parameter description it is best to mention to the user to add this environment variable. 
+
+Read more:
+* [Environment Variable Name]({{site.baseurl}}/2.0/reusing-config/#environment-variable-name)
+* [Best Practices]({{site.baseurl}}/2.0/orbs-best-practices/)
+
+## Environment Variables
+
+* Question: How can I require a user to add an environment variable?
+* Answer: Create a parameter for the environment variable name, even if it is a statically named environment variable the user _should not_ change, and assign it the correct default value. In the parameter description let the user know if this value should not be changed. In either event instruct the user where they can obtain their API key.
```
Comment (lokst) [inline jekyll/_cci2/orb-author-faq.md]: I was thinking the required env vars could be documented and in the orb there could be a check done for the presence of the env vars. For example:

```
if [ -z "$MY_ENV_VAR" ]; then
  echo "Error: please set the value of the MY_ENV_VAR environment variable"
  exit 1
fi
```

File: jekyll/_cci2/orb-author-faq.md [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+* [Best Practices]({{site.baseurl}}/2.0/orbs-best-practices/)
+
+## Supported Programming Languages
+
+* Question: What programming language can I write my orb in?
+* Answer: Orbs are packages of [CircleCI YAML config]({{site.baseurl}}/2.0/configuration-reference/) language. CircleCI config may contain commands that are issued against a given [Executor]({{site.baseurl}}/2.0/executor-intro/).
+
+CircleCI orbs package [CircleCI reusable config]({{site.baseurl}}/2.0/reusing-config/), such as [Commands]({{site.baseurl}}/2.0/reusing-config/#authoring-reusable-commands), which can execute within a given [Executor]({{site.baseurl}}/2.0/executor-intro/) defined by either the user if using a command within a custom job, or by the orb author if using a [Reusable Job]({{site.baseurl}}/2.0/orb-author-intro/#jobs). The environment within which your logic is running may influence your language descisions.
+
+**Bash**
```
Comment (lokst) [inline jekyll/_cci2/orb-author-faq.md]: Should we encourage compliance with POSIX shell language as much as possible?

File: jekyll/_cci2/orb-author-faq.md (hunk 4 above)
Comment (KyleTryon) [reply jekyll/_cci2/orb-author-faq.md]: I think we can consider that a general good practice we should add the new orbs-best-practices and maybe link back to here. I'm going to open that PR, and then reference that here.

https://github.com/circleci/circleci-docs/pull/3773

^ Will update this PR with a link once this is merged.

File: jekyll/_cci2/orb-author-faq.md (hunk 5 above)
Comment (KyleTryon) [reply jekyll/_cci2/orb-author-faq.md]: I think so yes. I prefer if _only_ POSIX shell language was used but do also want to address people's questions and concerns. Especially if they are writing orbs for themselves and don't intend to share.

Did you have any specific language you wanted to add? Maybe we should add a more explicit warning about bash being the most suitable for sharing?

File: jekyll/_cci2/orb-author-faq.md [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+Read more:
+* [Environment Variable Name]({{site.baseurl}}/2.0/reusing-config/#environment-variable-name)
+* [Best Practices]({{site.baseurl}}/2.0/orbs-best-practices/)
+
+## Supported Programming Languages
+
+* Question: What programming language can I write my orb in?
+* Answer: Orbs are packages of [CircleCI YAML config]({{site.baseurl}}/2.0/configuration-reference/) language. CircleCI config may contain commands that are issued against a given [Executor]({{site.baseurl}}/2.0/executor-intro/).
+
+CircleCI orbs package [CircleCI reusable config]({{site.baseurl}}/2.0/reusing-config/), such as [Commands]({{site.baseurl}}/2.0/reusing-config/#authoring-reusable-commands), which can execute within a given [Executor]({{site.baseurl}}/2.0/executor-intro/) defined by either the user if using a command within a custom job, or by the orb author if using a [Reusable Job]({{site.baseurl}}/2.0/orb-author-intro/#jobs). The environment within which your logic is running may influence your language decisions.
```
Comment (teesloane) [inline jekyll/_cci2/orb-author-faq.md]: I find the first line on `58` pretty similar to the ending of `line 56` ("CircleCI config may contain commands that are issued against a given [Executor]({{site.baseurl}}/2.0/executor-intro/).")

Can we remove one of these?

File: jekyll/_cci2/orb-author-faq.md [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+* Question: How can I require a user to add an environment variable?
+* Answer: Create a parameter for the environment variable name, even if it is a statically named environment variable the user _should not_ change, and assign it the correct default value. In the parameter description let the user know if this value should not be changed. In either event instruct the user where they can obtain their API key. 
+
+Read more:
+* [Environment Variable Name]({{site.baseurl}}/2.0/reusing-config/#environment-variable-name)
+* [Best Practices]({{site.baseurl}}/2.0/orbs-best-practices/)
+
+## Supported Programming Languages
+
+* Question: What programming language can I write my orb in?
```
Comment (teesloane) [inline jekyll/_cci2/orb-author-faq.md]: Can we generalize this question a bit more? I was a bit confused just by the use of the word `write` - in that I never imagined I could "write" an orb with bash, python, etc. Would something like

"What programming languages/tools can I use within my orb?"

And/or, maybe we separate this question into: 

- what language do I use to write an orb (a: cci yaml); &
- what programming languages tools can I use within an orb

Just a thought!

Comment (KyleTryon) [reply jekyll/_cci2/orb-author-faq.md]: How about this? https://github.com/circleci/circleci-docs/pull/3770/commits/45a331d8e60f8c5c868571b5d3fb551f56e73956

File: jekyll/_cci2/orb-author-faq.md (hunk 6 above)
Comment (KyleTryon) [reply jekyll/_cci2/orb-author-faq.md]: Removed from line 56.

File: jekyll/_cci2/orb-author-faq.md (hunk 4 above)
Comment (KyleTryon) [reply jekyll/_cci2/orb-author-faq.md]: Added: https://github.com/circleci/circleci-docs/pull/3770/commits/3394e3cfda2369251fee584fdb256b1741bc8788

File: jekyll/_cci2/orb-author-faq.md (hunk 5 above)
Comment (KyleTryon) [reply jekyll/_cci2/orb-author-faq.md]: mild mention here: https://github.com/circleci/circleci-docs/pull/3770/files#diff-06b742329a74ede7ed5da1da9ff31b23R63

Comment (lokst) [reply jekyll/_cci2/orb-author-faq.md]: @KyleTryon Thanks for adding the POSIX mention!

**Comment by KyleTryon** [comment]
@michelle-luna I believe we are ready to publish and merge now 👍 Whenever ya'll are ready to have it added.

**Pull request merged**